lib/fileset.nix

```diff
@@ -0,0 +1,993 @@
+          value = rhs;
+        }
+      ];
+    in
+    _create normalised.commonBase
+      (_intersectTree
+        (elemAt normalised.trees 0)
+        (elemAt normalised.trees 1)
+      );
+
+  /*
+  The file set containing all files that are in all the given sets.
+  Recursively, the elements are evaluated from left to right, only evaluating arguments on the right if necessary.
+
+  Type:
+    intersects :: [FileSet] -> FileSet
+  */
+  intersects = list:
+    let
+      annotated = imap0 (i: el: {
+        context = "element ${toString i} of the argument";
+        value = el;
+      }) list;
+
+      normalised = _normaliseBase "intersects" annotated;
+
+      tree = foldl' _intersectTree (head normalised.trees) (tail normalised.trees);
+    in
+    if ! isList list then
+      throw "lib.fileset.intersects: Expected argument to be a list, but got a ${typeOf list}."
+    else if length list == 0 then
+      throw "lib.fileset.intersects: Expected argument to be a list with at least one element, but it contains no elements."
+    else
+      _create normalised.commonBase tree;
+
+  /*
+  The file set containing all files that are in the first file set but not in the second.
+  Recursively, the second argument is evaluated first, only evaluating the first argument if necessary.
+
+      difference a b == a ∖ b
+
+  Type:
+    difference :: FileSet -> FileSet -> FileSet
+  */
+  difference = lhs: rhs:
+    let
+      normalised = _normaliseBase "difference" [
+        {
+          context = "first argument";
+          value = lhs;
+        }
+        {
+          context = "second argument";
+          value = rhs;
+        }
+      ];
+    in
+    _create normalised.commonBase
+      (_differenceTree normalised.commonBase
+        (elemAt normalised.trees 0)
+        (elemAt normalised.trees 1)
+      );
+
+  /*
+  Filter a file set to only contain files matching some predicate.
+
+  The predicate is called with an attribute set containing these attributes:
+
+  - `name`: The filename
+
+  - `type`: The type of the file, either "regular", "symlink" or "unknown"
+
+  - `ext`: The file extension or `null` if the file has none.
+
+    More formally:
+    - `ext` contains no `.`
+    - `.${ext}` is a suffix of the `name`
+
+  - Potentially other attributes in the future
+
+  Type:
+    fileFilter ::
+      ({
+        name :: String,
+        type :: String,
+        ext :: String | Null,
+        ...
+      } -> Bool)
+      -> FileSet
+      -> FileSet
+  */
+  fileFilter = predicate: maybeFileset:
+    let
+      fileset = _coerce "fileFilter" "second argument" maybeFileset;
+      recurse = focusPath: tree:
+        mapAttrs (name: subtree:
+          if isAttrs subtree || subtree == "directory" then
+            recurse (append focusPath name) subtree
+          else if
+            predicate {
+              inherit name;
+              type = subtree;
+              ext = mapNullable head (match ".*\\.(.*)" name);
```

Comment:
We'd get better orthogonality if we provided that as a library function, and an example of use in this function's documentation.